src/Equinox.MessageDb/MessageDbClient.fs

```diff
@@ -0,0 +1,126 @@
+                let time =
+                    DateTime.SpecifyKind(reader.GetDateTime(7), DateTimeKind.Utc)
+
+                let timestamp = DateTimeOffset(time)
+
+                let timelineEvent =
+                    TimelineEvent.Create(
+                        index = reader.GetInt64(0),
+                        eventType = reader.GetString(1),
+                        data = JsonSerializer.Deserialize<JsonElement>(reader.GetString(2)),
+                        meta = JsonSerializer.Deserialize<JsonElement>(meta),
+                        eventId = reader.GetGuid(4),
+                        ?correlationId =
+                            (if reader.IsDBNull(5) then
+                                 None
+                             else
+                                 Some(reader.GetString(5))),
+                        ?causationId =
+                            (if reader.IsDBNull(6) then
+                                 None
+                             else
+                                 Some(reader.GetString(6))),
+                        timestamp = timestamp
+                    )
+
+                events.Add(timelineEvent)
+                let! next = reader.ReadAsync(ct)
+                hasNext <- next
+
+            return events.ToArray()
+        }
+
+    member private _.PrepareWriteCommand
+        (cmd: NpgsqlBatchCommand)
+        (streamName: string)
+        version
+        (message: IEventData<JsonElement>)
+        =
+        cmd.CommandText <- "select 1 from write_message(@Id::text, @StreamName, @EventType, @Data, @Meta, @ExpectedVersion)"
+
+        cmd.Parameters.AddWithValue("Id", NpgsqlDbType.Uuid, message.EventId) |> ignore
+        cmd.Parameters.AddWithValue("StreamName", NpgsqlDbType.Text, streamName) |> ignore
+        cmd.Parameters.AddWithValue("EventType", NpgsqlDbType.Text, message.EventType) |> ignore
+        let data, meta = message.Data, message.Meta
+        cmd.Parameters.AddWithValue("Data", NpgsqlDbType.Jsonb, data.GetRawText()) |> ignore
+        let meta =
+            match meta.ValueKind with
+            | JsonValueKind.Null | JsonValueKind.Undefined -> "null"
+            | _ -> meta.GetRawText()
+        cmd.Parameters.AddWithValue("Meta", NpgsqlDbType.Jsonb, meta) |> ignore
+        cmd.Parameters.AddWithValue("ExpectedVersion", NpgsqlDbType.Bigint, version) |> ignore
+
+        cmd
+
+    member __.WriteMessages(streamName, events, version: int64, ct) =
+        task {
+            try
+                use! conn = source ()
+                use transaction = conn.BeginTransaction()
+                use batch = new NpgsqlBatch(conn, transaction)
+
+                let mutable expectedVersion = version
+
+                for event in events do
+                    batch.BatchCommands.Add(
+                        __.PrepareWriteCommand (NpgsqlBatchCommand()) streamName expectedVersion event
+                    )
+
+                    expectedVersion <- expectedVersion + 1L
```

Comment:
force input to be an array and use the `+ Array.length events` instead;

Reply:
This is happening while looping over the sequence. Each insert has the expected version of `index + previousVersion`

I thought this might be preferable to an `Array.mapi`

Reply:
Ha, I even missed that point because I had not seen such mutation in recent times ;)
In general I tend to favour expressing such things as a `for` also (and I deleted a suggestion to use an `iter`)
I think expressing it by calculating `expectedVersion + i` within an `iteri` would work well
perhaps make a helper
```
let prep i event = __.PrepareWriteCommand (NpgsqlBatchCommand()) streamName (expectedVersion+i) event
events |> Array.iteri prep
```
(it'll look cleaner when PrepareWriteCommand gets tupled args or becomes a function too)